### PR TITLE
Replacing all_tests.sh

### PR DESCRIPTION
**Pull request by bendk**
- Created tests.py to replace it.
- Added some new functionality targeted at developers who want to test
  their changes before submitting them for review.
- Moved the functionality from `all_rests_test.sh` into `tests.py`.
  Split it up into 2 parts: rust-tests and nss-bindings.
- Added modes to perform the other tests/linting/formatting that
  all_tests.sh was handling.
- Updated CircleCI config to use tests.py when possible.

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - Consider running `automation/all_tests.sh` to execute these tests locally.
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

**Comment by bendk** [comment]
Pushing my attempt to update `all_tests.sh`.  The main goals here were to:

- Provide devs with a single command that runs a reasonable set of tests/linters/formatters before they submit their code for review.  I tried to make this run in quick enough that devs would actually not feel pulled to skip these because they were too slow.
- Consolidate the testing/linting/formatting logic a bit.  Before it was split up over `all_tests.sh`, `all_clippy_checks.sh`, `all_rust_tests.sh`.

This breaks the asdev integration, but it should be easy to fix.

File: automation/tests.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+                                      default_features_only=True))
+
+    if not branch_changes.paths:
+        print('no changes found.')
+        return
+
+    if branch_changes.has_unstanged_changes():
+        subprocess.run(['git', 'status'])
+        print()
+        print('WARNING: unstanged changes in your branch:')
```
Comment (mhammond) [inline automation/tests.py:351]: typo - 'unstaged'

**Review by mhammond — COMMENTED** [review]
LGTM!

**Comment by bendk** [comment]
The check-dependencies step is not working, but I don't think that's related to these changes.  We talked about this on the call today and decided to try this one out.

**Pull request merged**